File: Tree.c
#include <cstdlib>
#include <cstdio>
#include <iostream>
#include <string.h>
#include "Tree.h"
using namespace std;

#ifndef TRUE
#define TRUE 1
#endif

#ifndef FALSE
#define FALSE 0
#endif

static const char ALLOCATE[] = " - Allocating]\n";
static const char DEALLOCATE[] = " has been deallocated]\n";
static const char AND[] = " and ";
static const char COMPARE[] = " - Comparing ";
static const char INSERT[] = " - Inserting ";
static const char TREE[] = "[Tree ";

/**
 * The following struct contains functions associated with a tree node
 * such as constructor, destructor and write functions
 *
 * @author Mark Zeng(cs12fa19dm)
 */
struct TNode {
	Base * data;
	TNode * left, * right, *parent;
	static long occupancy;
        unsigned long hasBeenDeleted;

	// left child's height - right child's height
	long balance;

	// 1 + height of tallest child, or 0 for leaf
	long height;

TNode (Base * element) : data (element), left (0), right (0),
	parent (0), hasBeenDeleted (FALSE), balance (0), height(0) 
/***************************************************************************
% Routine Name : TNode (public)
% File :         Tree.c
% 
% Description : This function will allocates and initializes the memory 
%               associated with the TNode object.
%
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% element           the data stored in the node
***************************************************************************/
{
	occupancy++;
}

~TNode (void) 
/*************************************************************************** 
% Routine Name : ~TNode  (public) 
% File :         Tree.c 
%  
% Description :  deallocates memory associated with the Tree Node.
***************************************************************************/  
{
	delete(data);
	occupancy--;
}
        
ostream & Write (ostream & stream) const {
                stream << "at height:  " << height << "  with balance:  "
			<< balance << "  ";
		return data->Write (stream) << "\n";
        }
	
ostream & Write_AllTNodes (ostream & stream) const {
		if (left)
			left->Write_AllTNodes (stream);
		if (!hasBeenDeleted)
			Write (stream);
		if (right)
			right->Write_AllTNodes (stream);

		return stream;
	}

void delete_AllTNodes (void) 
/*************************************************************************** 
% Routine Name : delete_AllTNodes  (public) 
% File :         Tree.c 
%  
% Description :  use post-order traversal to delete all the nodes
***************************************************************************/  
{
	if(left)
		left -> delete_AllTNodes();
	
	if(right)
		right -> delete_AllTNodes();
         
	delete(this);
}

};

// initialization of static data fields of Tree
long TNode :: occupancy = 0;
bool Tree :: debug_on = 0;

void Tree :: Set_Debug (bool option) {
/*************************************************************************** 
% Routine Name : Tree :: Set_Debug (public) 
% File :         Tree.c
%  
% Description :  This function sets debug mode on or off 
% 
% Parameters descriptions : 
%  
% name               description 
% ------------------ ------------------------------------------------------ 
% option             true should set debug mode on, false should set debug 
%                    mode off. 
***************************************************************************/  
	debug_on = option;
}

Tree :: Tree (const char * name) : root (0), tree_name (strdup (name))
/***************************************************************************
% Routine Name : Tree :: Tree  (public)
% File :         Tree.c
% 
% Description :  Initializes root pointer to NULL, and occupancy to 0.
***************************************************************************/

{
	TNode :: occupancy = 0;
	if(debug_on){
		cerr << TREE << tree_name << ALLOCATE ;
	}
}

Tree :: ~Tree (void)
/***************************************************************************
% Routine Name : Tree :: ~Tree  (public)
% File :         Tree.c
% 
% Description :  deallocates memory associated with the Tree.  It
%                will also delete all the memory of the elements within
%                the table.
***************************************************************************/

{
	if(debug_on){
		cerr << TREE << tree_name << DEALLOCATE;
	}
	/*delete node*/
	TNode :: occupancy = 0;
	if(root != NULL){
	root -> delete_AllTNodes();
	}
	/*delete tree*/
	free((void*)tree_name);
}	/* end: ~Tree */

unsigned long Tree :: IsEmpty () const 
/***************************************************************************
% Routine Name : Tree :: IsEmpty  (public)
% File :         Tree.c
% 
% Description :  checks if the Tree is Empty
% 
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% <return>           1 if empty, 0 if not
***************************************************************************/
{
	return (TNode :: occupancy)? 0 : 1;
}

void updateHeight (TNode * current)
/***************************************************************************
% Routine Name : Tree :: updateHeight  (private)
% File :         Tree.c
% 
% Description :  this function updates height of each node after insert
***************************************************************************/
{
	/*check if height realy needs to be updated*/
	if(current -> left != NULL && current -> right != NULL){
		return ;
	}
	/*update the height*/
	while(true){
		current -> height++;
		/*check root*/
		if(current -> parent == NULL)
			break;
		current = current -> parent;
	}
}

void updateBalance (TNode * current)
/***************************************************************************
% Routine Name : Tree :: updateBalance  (private)
% File :         Tree.c
% 
% Description :  this function updates balance of each node after insert
***************************************************************************/
{
	long leftHeight; /*the height of left node*/
	long rightHeight;/*the height of right node*/
		
	/*check balance*/
	while(true) {
		/*check left*/
		if(current->left == NULL){
			leftHeight = -1;
		}
		else{
			leftHeight = current -> left -> height;
		}
		/*check right*/
		if(current -> right == NULL){
			rightHeight = -1;
		}
		else{
			rightHeight = current -> right -> height;
		}
                current -> balance = leftHeight - rightHeight;
		/*check root*/
		if(current -> parent == NULL)
			break;
		current = current -> parent;
	}
}
unsigned long Tree :: Insert (Base * element) 
/***************************************************************************
% Routine Name : Tree :: Insert (public)
% File :         Tree.c
% 
% Description : Inserts the element into the binary tree. Returns true or 
%		false indicating success of insertion.
% 
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% element            element to be inserted
% <return>           1 if scucess, 0 if not
***************************************************************************/
{
	TNode * current;   /*used to keep track of the current node*/
	
	/*check if root exist*/
	if(root == NULL){
		if(debug_on){
			cerr << TREE << tree_name << INSERT << 
					(const char *)*element << "]" << endl;
		}
		root = new TNode(element);
               	return true;
	}
	current = root;
	while(true){	
		if(debug_on){
			cerr << TREE << tree_name << COMPARE <<  
					(const char *)*element << AND << 
				       	(const char *)*current -> data << "]" <<  
					endl;
		}
		/*check to go right or left*/
		if(*element > *current -> data){
			/*check if right exists*/
			if(current->right == NULL){
				current->right = new TNode(element);
				current->right->parent = current;
				updateHeight(current);
				updateBalance(current);
				if(debug_on){
				cerr << TREE << tree_name << INSERT << 
					(const char *)*element << "]" << endl;
				}
				return true;
			}
			/*go right*/
			else{
				current = current->right;
				continue;
			}
		}
		else{
			/*check if left exist*/
			if(current->left == NULL){
				current->left = new TNode(element);
				current->left->parent = current;
				updateHeight(current);
				updateBalance(current);
				if(debug_on){
				cerr << TREE << tree_name << INSERT << 
					(const char *)*element << "]" << endl;
				}
				return true;
			}
			/*go left*/
			else{
				current = current->left;
				continue;
			}
		}
	}
}

const Base * Tree :: Lookup (const Base * element) const 
/***************************************************************************
% Routine Name : Tree :: Lookup (public)
% File :         Tree.c
% 
% Description : Look up the element in the binary tree. Returns pointer to data
%		or NULL
% 
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% element            element to be looked up
% <return>           pointer to data if found, null if not
***************************************************************************/
{
	TNode * current;   /*used to keep track of the current node*/
	
	/*check if root exist*/
	if(root == NULL){
		return NULL;
	}
	current = root;
	
	/*checking down the tree*/
	while(true){	
		if(debug_on){
			cerr << TREE << tree_name << COMPARE <<  
					(const char *)*element << AND << 
				       	(const char *)*current -> data << "]" <<  
					endl;
		}
		/*go right?*/
		if(*element > *current->data){
			/*check if right exists*/
			if(current->right == NULL){
				return NULL;
			}
			/*go right*/
			else{
				current = current->right;
				continue;
			}
		}
		/*found*/
		else if(*element == *current->data){
			/*deleted?*/
			if(current->hasBeenDeleted == false){
				return current->data;
			}
			else{
				return NULL;
			}
		}
		else{
			/*check if left exists*/
			if(current->left == NULL){
				return NULL;
			}
			/*go left*/
			else{
				current = current->left;
				continue;
			}
		}
	}
}

Base * Tree :: Remove (const Base * element) 
/***************************************************************************
% Routine Name : Tree :: Remove (public)
% File :         Tree.c
% 
% Description : remove the matching data in the binary tree. Returns pointer to
%		data or NULL
% 
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% element            element to be removed
% <return>           pointer to data if found, null if not
***************************************************************************/
{
	TNode * current;   /*used to keep track of the current node*/
	
	/*check if root exist*/
	if(root == NULL){
		return NULL;
	}
	current = root;
	
	/*checking down the tree*/
	while(true){	
		if(debug_on){
			cerr << TREE << tree_name << COMPARE <<  
					(const char *)*element << AND << 
				       	(const char *)*current -> data << "]" <<  
					endl;
		}
		/*go right?*/
		if(*element > *current->data){
			/*check if right exists*/
			if(current->right == NULL){
				return NULL;
			}
			/*go right*/
			else{
				current = current->right;
				continue;
			}
		}
		/*found*/
		else if(*element == *current->data){
			/*deleted?*/
			if(current->hasBeenDeleted == false){
				current->hasBeenDeleted = true;
				TNode :: occupancy--;
				return current->data;
			}
			else{
				return NULL;
			}
		}
		else{
			/*check if left exists*/
			if(current->left == NULL){
				return NULL;
			}
			/*go left*/
			else{
				current = current->left;
				continue;
			}
		}
	}
}

ostream & Tree :: Write (ostream & stream) const
/***************************************************************************
% Routine Name : Tree :: Write (public)
% File :         Tree.c
% 
% Description : This function will output the contents of the Tree table
%               to the stream specificed by the caller.  The stream could be
%               cerr, cout, or any other valid stream.
%
% Parameters descriptions :
% 
% name               description
% ------------------ ------------------------------------------------------
% stream             A reference to the output stream.
% <return>           A reference to the output stream.
***************************************************************************/

{
        stream << "Tree " << tree_name << ":\n"
		<< "occupancy is " << TNode :: occupancy << " elements.\n";

	return (root) ? root->Write_AllTNodes (stream) : stream;
}       /* end: Write */

File: Tree.java
/**
 * this class implements a binary tree sturcture, with methods
 * including insert, lookup , remove and write.
 *
 * @author Mark Zeng(cs12fa19dm)
 */
public class Tree<Whatever extends Base> extends Base {

        /* data fields */
        private TNode root;
        private long occupancy; 
        private String treeName;

        /* debug flag */
        private static boolean debug;

        /* debug messages */
        private static final String ALLOCATE = " - Allocating]\n";
        private static final String AND = " and ";
        private static final String COMPARE = " - Comparing ";
        private static final String INSERT = " - Inserting ";
        private static final String TREE = "[Tree ";

        /**
	 * Constructer
	 * Allocates and initializes the memory associated with
	 * the Tree object.
	 *
	 * @param takes in a string name
	 */
        public Tree (String name) {
		treeName = name;
		occupancy = 0;
		root = null;
		if(debug){
			System.err.print(TREE + name + ALLOCATE);
		}	
        }
   	
	/**
         * This method set debug state to on
         */
        public static void debugOff () {

                debug = false;
        }
        /**
         * This method set debug state to on
         */
      	public static void debugOn () {

                debug = true;
        }

        /**
         * Returns the tree's name
         * 
	 * @return name of the tree
         */
        public String getName() {
                return treeName;
        }
	
	/**
	 * Return if the tree is empty
	 *
	 * @return true if empty, false if otherwise
	 */
	public boolean isEmpty () {
		if(occupancy == 0){
			return true;
		}
		else{
			return false;
		}
	}
        
	/**
	 * This method updates height of each node after insert
	 * 
	 * @param The Node that needs to be updated
	 */
	public void updateHeight(TNode current) {
		/*check if height realy needs to be updated*/
		if(current.left != null && current.right != null){
			return ;
		}
		/*update the height*/
		while(true){
			current.height++;
			/*check root*/
			if(current.parent == null)
				break;
			current = current.parent;
		}
	}

	/**
	 * This method updates balance of each node after insert
	 *
	 * @param The Node needs to be updated
	 */
	public void updateBalance(TNode current) {
		long leftHeight; /*the height of left node*/
		long rightHeight;/*the height of right node*/
		
		/*check balance*/
		while(true) {
			/*check left*/
			if(current.left == null){
				leftHeight = -1;
			}
			else{
				leftHeight = current.left.height;
			}
			/*check right*/
			if(current.right == null){
				rightHeight = -1;
			}
			else{
				rightHeight = current.right.height;
			}
                        current.balance = leftHeight - rightHeight;
			/*check root*/
			if(current.parent == null)
				break;
			current = current.parent;
		}
	}
		
	/**
	 * This method inserts the element into the binary tree.returns true or
	 * false indicating seccess of insertion
	 *
	 * @param element is the complete element to be inserted
	 * @return true or false indicating seccess of insertion
	 */
        public boolean insert (Whatever element) {
                TNode current;   /*used to keep track of the current node*/
	
		/*check if root exist*/
		if(root == null){
			if(debug){
				System.err.println(TREE + treeName + INSERT +
						element.getName() + "]");
			}
			root = new TNode(element);
                	return true;
		}
		current = root;
		while(true){	
			if(debug){
				System.err.println(TREE + treeName + COMPARE +
						element.getName() + AND +
					       	current.data.getName() + "]");
			}
			/*check to go right or left*/
			if(element.isGreaterThan(current.data)){
				/*check if right exists*/
				if(current.right == null){
					current.right = new TNode(element);
					current.right.parent = current;
					updateHeight(current);
					updateBalance(current);
					if(debug){
						System.err.println(TREE + 
						treeName + INSERT +
						element.getName() + "]");
					}
					return true;
				}
				/*go right*/
				else{
					current = current.right;
					continue;
				}
			}
			else{
				/*check if left exist*/
				if(current.left == null){
					current.left = new TNode(element);
					current.left.parent = current;
					updateHeight(current);
					updateBalance(current);
					if(debug){
						System.err.println(TREE + 
						treeName + INSERT +
						element.getName() + "]");
					}
					return true;
				}
				/*go left*/
				else{
					current = current.left;
					continue;
				}
			}
		}
        }
        
	/**
	 * this method Looks up the matching data in the binary tree. 
	 * Returns a pointer to the data if found, null otherwise. 
	 *
	 * @param the element to be looked up
	 * @return a pointer to the data if found, null otherwise
	 */
        public Whatever lookup (Whatever element) {
		TNode current; /*used to keep track of the current node*/
                
		/*check if root exists*/
		if(root == null){
			return null;
		}
                current = root;
		
		/*checking down the tree*/
		while(true){
			if(debug){
				System.err.println(TREE + treeName + COMPARE +
						element.getName() + AND +
						current.data.getName() + "]");
			}
			/*go right?*/
			if(element.isGreaterThan(current.data)){
				/*find nothing*/
				if(current.right == null){
					return null;
				}
				else{
					current = current.right;
					continue;
				}
			}
			/*found*/
			else if(element.equals(current.data)){
				/*deleted?*/
				if(current.hasBeenDeleted == false){
					return current.data;
				}
				else{
					return null;
				}
			}
			/*go left?*/
			else{
				/*find nothing*/
				if(current.left == null){
					return null;
				}
				else{
					current = current.left;
					continue;
				}
			}
		}
        }
	
	/**
	 * this method removes the matching data in the binary tree. 
	 * Returns a pointer to the data if found, null otherwise. 
	 *
	 * @param the element to be looked up
	 * @return a pointer to the data if found, null otherwise
	 */
        public Whatever remove (Whatever element) {
 		TNode current; /*used to keep track of the current node*/
                
		/*check if root exists*/
		if(root == null){
			return null;
		}
                current = root;
		
		/*checking down the tree*/
		while(true){
			if(debug){
				System.err.println(TREE + treeName + COMPARE +
						element.getName() + AND +
						current.data.getName() + "]");
			}
			/*go right?*/
			if(element.isGreaterThan(current.data)){
				/*find nothing*/
				if(current.right == null){
					return null;
				}
				else{
					current = current.right;
					continue;
				}
			}
			/*found*/
			else if(element.equals(current.data)){
				/*deleted?*/
				if(current.hasBeenDeleted == false){
					current.hasBeenDeleted = true;
					occupancy--;
					return current.data;
				}
				else{
					return null;
				}
			}
			/*go left?*/
			else{
				/*find nothing*/
				if(current.left == null){
					return null;
				}
				else{
					current = current.left;
					continue;
				}
			}
		}
        }

        /**
         * Creates a string representation of this tree. This method first
         * adds the general information of this tree, then calls the
         * recursive TNode function to add all nodes to the return string 
         *
         * @return  String representation of this tree 
         */
        public String toString () {
                String string = "Tree " + treeName + ":\noccupancy is ";
                string += occupancy + " elements.";

                if(root != null)
                        string += root.writeAllTNodes();

                return string;
        }

	/**
	 * this class implements basic methods of a treenode, including
	 * a constructor , toSrting method and writeAllNodes mathods.
	 *
	 * @author Mark Zeng(cs12fa19dm)
	 */
        private class TNode {

                public Whatever data;
                public TNode left, right, parent;
                public boolean hasBeenDeleted;

                /* left child's height - right child's height */
                public long balance;
                /* 1 + height of tallest child, or 0 for leaf */
                public long height;
                
		/**
		 * Constructor
		 * allocate and initializes the memory associated with
		 * the TNode object
		 * 
		 * @param element is the data stored in the TNode
		 */
                public TNode (Whatever element) {
			data = element;
			left = null;
			right = null;
			parent = null;
			hasBeenDeleted = false;
			balance = 0;
			height = 0;
			occupancy++;
                }

                /**
                 * Creates a string representation of this node. Information
                 * to be printed includes this node's height, its balance,
                 * and the data its storing.
                 *
                 * @return  String representation of this node 
                 */

                public String toString () {
                        return "at height:  " + height + "  with balance: " +
                                balance + "  " + data;
                }

                /**
                 * Writes all TNodes to the String representation field. 
                 * This recursive method performs an in-order
                 * traversal of the entire tree to print all nodes in
                 * sorted order, as determined by the keys stored in each
                 * node. To print itself, the current node will append to
                 * tree's String field.
                 */
                public String writeAllTNodes () {
                        String string = "";
                        if (left != null)
                                string += left.writeAllTNodes ();
                        if (!hasBeenDeleted) 
                                string += "\n" + this;          
                        if (right != null)
                                string += right.writeAllTNodes ();

                        return string;
                }
        }
}
